cpp/examples/parquet_io/parquet_io_multithreaded.cpp

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "common.hpp"
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/statistics_resource_adaptor.hpp>
+
+#include <fmt/chrono.h>
+
+#include <filesystem>
+
+/**
+ * @file parquet_io_multithreaded.cpp
+ * @brief Demonstrates usage of the libcudf APIs to read and write
+ * parquet file format with different encodings and compression types
+ * using multiple threads.
+ *
+ * The following encoding and compression ztypes are demonstrated:
+ * Encoding Types: DEFAULT, DICTIONARY, PLAIN, DELTA_BINARY_PACKED,
+ *                 DELTA_LENGTH_BYTE_ARRAY, DELTA_BYTE_ARRAY
+ *
+ * Compression Types: NONE, AUTO, SNAPPY, LZ4, ZSTD
+ *
+ */
+
+using table_t = std::unique_ptr<cudf::table>;
+
+struct read_fn {
+  std::vector<std::string> const& input_files;
+  std::vector<table_t>& tables;
+  int const thread_id;
+  int const thread_count;
+  rmm::cuda_stream_view stream;
+
+  void operator()()
+  {
+    std::vector<table_t> tables_this_thread;
+    for (auto curr_file_idx = thread_id; curr_file_idx < input_files.size();
+         curr_file_idx += thread_count) {
+      auto const source_info = cudf::io::source_info(input_files[curr_file_idx]);
+      auto builder           = cudf::io::parquet_reader_options::builder(source_info);
+      auto const options     = builder.build();
+      tables_this_thread.push_back(cudf::io::read_parquet(options, stream).tbl);
+    }
+
+    // Concatenate all tables read by this thread.
+    auto table = std::move(tables_this_thread[0]);
+    std::for_each(tables_this_thread.begin() + 1, tables_this_thread.end(), [&](auto& tbl) {
+      std::vector<cudf::table_view> const table_views{table->view(), tbl->view()};
+      table = cudf::concatenate(table_views, stream);
+    });
```

Comment:
This concatenate step makes it so I can't randomly assign files to each thread. I think I prefer stripping this example down to just be a multi-thread reader, for the purposes of better studying read_parquet pipelining.

Reply:
All done. The example will now only read unless optionally specified to write.